Make "What's new in C# 6" interactive (#9551)
dotnet/docs: docs/csharp/whats-new/csharp-6.md

```csharp
// preferred in C# 6:
this.SomethingHappened?.Invoke(this, eventArgs);
```

Ensuring that the left side is evaluated only once also enables you to use any expression, including method calls, on the left side of the `?.`

## String interpolation

With C# 6, the new [string interpolation](../language-reference/tokens/interpolated.md) feature enables you to embed the expressions in a string. Simply preface the string with `$`and use expressions between `{` and `}` instead of ordinals:

[!code-csharp[stringInterpolation](../../../samples/snippets/csharp/new-in-6/newcode.cs#FullNameExpressionMember)]

This example uses properties for the substituted expressions. You can use any expression. For example, you could compute a student's grade point average as part of the interpolation:

[!code-csharp[stringInterpolationFormat](../../../samples/snippets/csharp/new-in-6/newcode.cs#stringInterpolationFormat)]

The preceding line of code formats the value for `Grades.Average()` as a floating-point number with two decimal places.

> Often you may need to format the string produced using a specific culture. You use the fact that the object produced by a string interpolation can be implicitly converted to <xref:System.FormattableString?displayProperty=nameWithType>. The <xref:System.FormattableString> instance contains the composite format string, and the results of evaluating the expressions before converting them to strings. Use the <xref:System.FormattableString.ToString(System.IFormatProvider)> method to specify the culture when formatting a string. For example, the following example produces a string using German culture. (It uses the ',' character for the decimal separator, and the '.' character as the thousands separator.)

Reviewer: ```suggestion
Often, you may need to format the string produced using a specific culture. You use the fact that the object produced by a string interpolation can be implicitly converted to <xref:System.FormattableString?displayProperty=nameWithType>. The <xref:System.FormattableString> instance contains the composite format string and the results of evaluating the expressions before converting them to strings. Use the <xref:System.FormattableString.ToString(System.IFormatProvider)?displayProperty=nameWithType> method to specify the culture when formatting a string. The following example produces a string using the German (de-DE) culture. (By default, the German culture uses the ',' character for the decimal separator, and the '.' character as the thousands separator.)
```

```csharp
FormattableString str = $"Average grade is {s.Grades.Average()}";
var gradeStr = str.ToString(new System.Globalization.CultureInfo("de-DE"));
```

> To get started with string interpolation, check the [String interpolation in C#](../tutorials/intro-to-csharp/interpolated-strings.yml) interactive tutorial, the [String interpolation](../language-reference/tokens/interpolated.md) article, and the [String interpolation in C#](../tutorials/string-interpolation.md) tutorial.

Reviewer: ```suggestion
To get started with string interpolation, see the [String interpolation in C#](../tutorials/intro-to-csharp/interpolated-strings.yml) interactive tutorial, the [String interpolation](../language-reference/tokens/interpolated.md) article, and the [String interpolation in C#](../tutorials/string-interpolation.md) tutorial.
```

## Exception filters

*Exception Filters* are clauses that determine when a given catch clause should be applied. If the expression used for an exception filter evaluates to `true`, the catch clause performs its normal processing on an exception. If the
expression evaluates to `false`, then the `catch` clause is skipped. One use is to examine information about an exception to determine if a `catch` clause can process the exception:

[!code-csharp[ExceptionFilter](../../../samples/snippets/csharp/new-in-6/NetworkClient.cs#ExceptionFilter)]

## The `nameof` expression

The `nameof` expression evaluates to the name of a symbol. It's a great way to get tools working whenever you need the name of a variable, a property, or a member field. One of the most common uses for `nameof` is to provide the name of a symbol that caused an exception:

[!code-csharp[nameof](../../../samples/snippets/csharp/new-in-6/NewCode.cs#UsingStaticString)]

Another use is with XAML-based applications that implement the `INotifyPropertyChanged` interface:

[!code-csharp[nameofNotify](../../../samples/snippets/csharp/new-in-6/viewmodel.cs#nameofNotify)]

## Await in Catch and Finally blocks

> C# 5 had several limitations around where you could place `await` expressions. You can now use `await` in `catch` or `finally` expressions.  With C# 6, you can also await in catch expressions. This is most often used with logging scenarios:

Reviewer: This is confusing. In C# 5, can you await in a finally, as this seems to imply? The mention of limitations in C# 5 also isn't particularly relevant, since it doesnt list the limitations. I've suggested some changes.

> C# 5 had several limitations around where you could place `await` expressions. You can now use `await` in `catch` or `finally` expressions.  With C# 6, you can also await in catch expressions. This is most often used with logging scenarios:

Reviewer: ```suggestion
With C# 5, you can now use `await` in `catch` or `finally` expressions.  This is most often used with logging scenarios:
```

[!code-csharp[AwaitFinally](../../../samples/snippets/csharp/new-in-6/NetworkClient.cs#AwaitFinally)]

The implementation details for adding `await` support inside `catch` and `finally` clauses ensure that the behavior is consistent with the behavior for synchronous code. When code executed in a `catch` or `finally` clause throws, execution looks for a suitable `catch` clause in the next surrounding block. If there was a current exception, that exception is lost. The same happens with awaited expressions in `catch` and `finally` clauses: a suitable `catch` is searched for, and the current exception, if any, is lost.  

> [!NOTE]
> This behavior is the reason it's recommended to write `catch` and `finally` clauses carefully, to avoid introducing new exceptions.

> ## Index initializers

Reviewer 2: I think the problem is is just with the refactor of info. This section is **Index Initializers** but you talk about **Collection Initializers** by referring to braces and the `Add` method. Also, the first code sample shows using a `List` type which isn't using index initializers.

Reviewer 2: OK just to clarify, this is pointing to the `## Index initializers` line and below. Not the change above that :D

> *Index Initializers* is one of two features that make collection initializers more consistent with index usage. In earlier releases of C#, you could use *collection initializers* only with sequence style collections, including <xref:System.Collections.Generic.Dictionary%602> by adding braces around key and value pairs:

Reviewer 3: @BillWagner I'm afraid such a presentation of index initializers might be understood in a limited way: only in the context of collections. Also mentioning of *collection initializers* doesn't help, because ultimately the feature is about *object initializers*. 

Let's take the [C# 5 specification](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-334.pdf) and check the "12.7.11.3 Object initializers" section (page 154). It starts with the following sentence:
>An _**object initializer**_ specifies values for zero or more fields or properties of an object.

Then, let's check the [corresponding section in the C# 6 specification draft](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#object-initializers). It starts with:
>An _**object initializer**_ specifies values for zero or more fields, properties or indexed elements of an object.

So, it's an update of the *object initializer* feature, and not of the *collection initializers* as one might interpret the following sentence:
>*Index Initializers* is one of two features that make collection initializers more consistent with index usage.

What makes it look like it's about collections is the fact that indexers usually have "collection semantics" with the `Dictionary` class being the most popular example of a class with an indexer. So, an upgrade of an object initializer feature mostly helps to initialize collections with indexers.

Here, I agree with [the comment](https://github.com/dotnet/docs/issues/5169#issuecomment-392382085) of @drewjcooper in the original [issue](https://github.com/dotnet/docs/issues/5169).

Author: I agree as well.  However, I think that goes deeper than I think we should on the "what's new" article.

I'm updating this PR to include that discussion in the document on [object initializers](https://docs.microsoft.com/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers).  Then, link to it from here.

Reviewer 3: @BillWagner oh...! That linked article about object initializers is indeed outdated. It misses indexers (starting with C# 6):
>Object initializers let you assign values to any accessible fields or properties of an object at creation time ...

Though, again, indexers are in the section about collections:
>You can specify indexed elements if the collection supports indexing.

> *Index Initializers* is one of two features that make collection initializers more consistent with index usage. In earlier releases of C#, you could use *collection initializers* with sequence style collections, including <xref:System.Collections.Generic.Dictionary%602> by adding braces around key and value pairs:

Reviewer: ```suggestion
*Index Initializers* is one of two features that make collection initializers more consistent with index usage. In earlier releases of C#, you could use *collection initializers* with sequence style collections, including <xref:System.Collections.Generic.Dictionary%602>, by adding braces around key and value pairs:
```

[!code-csharp[ListInitializer](../../../samples/snippets/csharp/new-in-6/initializers.cs#ListInitializer)]

> Now, you can use them with <xref:System.Collections.Generic.Dictionary%602> collections and other types where the accessible `Add` method accepts more than one argument. The new syntax supports assignment using an index into the collection:

Reviewer 3: I think, "similar types" here means "types with indexers". The following code:

```csharp
using System;
using System.Collections;

class Program
{
    static void Main(string[] args)
    {
        var f = new Foo
            {
                { "foo", "bar" },
                { 1, '1', "1" },
                { 2, '4', "8" },
                { 3, '9', "27" }
            };
    }
}

class Foo : IEnumerable
{
    public void Add(string s1, string s2) { Console.WriteLine("Two args"); }
    public void Add(int x, char y, string z) { Console.WriteLine("Three args"); }
    IEnumerator IEnumerable.GetEnumerator() { throw new NotImplementedException(); }
}
```

compiles and runs with the "C# 5" as the selected language version. Thanks to @drewjcooper for [the example](https://github.com/dotnet/docs/issues/5169#issuecomment-391895482). (I've updated it not to use C# 6 features.)

So, the support of the `Add` methods with more than one argument is not a new feature of C# 6. As far as I can tell from the changes in the [Collection initializers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#collection-initializers) section between C# 5 and C# 6, the only change is that the `Add` method might be an extension method (but it's already covered in the next section of this document).

The citation from the C# 5 spec (page 157):
>  For each specified element in order, the collection initializer invokes an `Add` method on the target object with the expression list of the element initializer as argument list,...

The element initializer might be an expression list (as `{ "foo", "bar" }` in the example), which is given as an argument list to an `Add` method.

Reviewer 2: If this is index initialization, does that rely on `Add`?

Author: No, object initializers using the indexer do not also require an `Add` method.

Reviewer 3: @BillWagner now, I understand better what confuses me in the following sentence:

>Now, you can use them with <xref:System.Collections.Generic.Dictionary%602> collections and other types where the accessible `Add` method accepts more than one argument.

First, I was wrong about what is behind "them" in "...you can use them with...". I thought "them" refers to index initializers. However, it refers to "braces around key and value pairs" from the paragraph that is above the code block. Is that correct understanding? I don't know what others would see behind "them", though. So, maybe no changes here.

Second, the very start of the sentence might be misleading:
>Now, you can...

Given that the topic is about what's new, I've read the sentence in the following way:
> Starting with C# 6, you can...

Starting with "now" might imply that that was not possible before.

I think the "now" might be dropped.

What do you think of the following version?
>You can use them with <xref:System.Collections.Generic.Dictionary%602> collections and other types where the accessible `Add` method accepts more than one argument.

I don't know how to replace "them" (if necessary), but dropping "now" wouldn't make the sentence worse.

> Now, you can use them with <xref:System.Collections.Generic.Dictionary%602> collections and other types where the accessible `Add` method accepts more than one argument. The new syntax supports assignment using an index into the collection:

Reviewer 3: The rationale behind this suggestion is in the [comment](https://github.com/dotnet/docs/pull/9551#discussion_r245432130)

```suggestion
You can use them with <xref:System.Collections.Generic.Dictionary%602> collections and other types where the accessible `Add` method accepts more than one argument. The new syntax supports assignment using an index into the collection:
```
